e2e/vaultcompat/vaultcompat_test.go

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package vaultcompat
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/go-cleanhttp"
+	"github.com/hashicorp/go-set"
+	"github.com/hashicorp/go-version"
+	nomadapi "github.com/hashicorp/nomad/api"
+	"github.com/hashicorp/nomad/testutil"
+	vaultapi "github.com/hashicorp/vault/api"
+	"github.com/shoenig/test/must"
+	"github.com/shoenig/test/wait"
+)
+
+const (
+	binDir  = "vault-bins"
+	envGate = "NOMAD_E2E_VAULTCOMPAT"
+)
+
+func TestVaultCompat(t *testing.T) {
+	if os.Getenv(envGate) != "1" {
+		t.Skip(envGate + " is not set; skipping")
+	}
+	t.Run("testVaultVersions", testVaultVersions)
+}
+
+func testVaultVersions(t *testing.T) {
+	versions := scanVaultVersions(t, getMinimumVersion(t))
+	versions.ForEach(func(b build) bool {
+		downloadVaultBuild(t, b)
+		testVaultBuild(t, b)
+		return true
+	})
+}
+
+func testVaultBuild(t *testing.T, b build) {
+	t.Run("vault("+b.Version+")", func(t *testing.T) {
+		vStop, vc := startVault(t, b)
+		defer vStop()
+		setupVault(t, vc)
+
+		nStop, nc := startNomad(t, vc)
+		defer nStop()
+		runCatJob(t, nc)
+
+		// give nomad and vault time to stop
+		defer func() { time.Sleep(5 * time.Second) }()
+	})
+}
+
+func runCatJob(t *testing.T, nc *nomadapi.Client) {
+	b, err := os.ReadFile("input/cat.hcl")
+	must.NoError(t, err)
+
+	jobs := nc.Jobs()
+	job, err := jobs.ParseHCL(string(b), true)
+	must.NoError(t, err, must.Sprint("failed to parse job HCL"))
+
+	_, _, err = jobs.Register(job, nil)
+	must.NoError(t, err, must.Sprint("failed to register job"))
+
+	must.Wait(t, wait.InitialSuccess(
+		wait.ErrorFunc(func() error {
+			allocs, _, err := jobs.Allocations(*job.ID, false, nil)
+			if err != nil {
+				return err
+			}
+			if n := len(allocs); n != 1 {
+				return fmt.Errorf("expected 1 alloc, got %d", n)
+			}
+			if s := allocs[0].ClientStatus; s != "complete" {
+				return fmt.Errorf("expected alloc status complete, got %s", s)
+			}
+			return nil
+		}),
+		wait.Timeout(20*time.Second),
+		wait.Gap(1*time.Second),
+	))
+
+	t.Log("success running cat job")
```

Comment:
this all hinges on `cat "${NOMAD_SECRETS_DIR}/vault_token"` exiting `0` -- not `1` "no such file" or other -- to result in "complete" ClientStatus instead of "failed" or perpetually-"pending" right?  and that exercises a bulk of the machinery to produce a vault token for the task.

is there any additional value in checking the (shape of) contents of that file, or the env var?  or confirming that the token can be used with an actual `vault token lookup` or similar?

Reply:
Yeah I had the same thought - for now this test just makes the same assertion as the code it replaces (that `cat`ing the token succeeds).

Let me make a followup ticket with your idea, it's probably worth doing but it's a good chunk more work than just cleaning things up. Using the Nomad API to gather logs from a specific task requires a lot of boilerplate.

Reply:
fwiw, a `vault token lookup` should exit `0` if there's a valid `VAULT_TOKEN` (and vault is reachable), and non-0 (`2`, apparently) if the token is invalid.  so adding that one more command to the "cat" job might be relatively low hanging fruit without having to check any log output.